File: find_error.py
#!/usr/bin/env python

import json
import argparse
from web3 import Web3

def generate_selector(signature):
    return Web3.keccak(text=signature).hex()[:10]

def find_matching_error(abi, error_selector_to_match):
    for item in abi:
        if item['type'] == 'error':
            # Generate the error signature
            error_signature = item['name'] + '(' + ','.join([input['type'] for input in item['inputs']]) + ')'
            # Generate the selector
            selector = generate_selector(error_signature)
            print(f"found error type: {item['name']} with selector: {selector}")

            # Compare with the known error selector
            if selector.startswith(error_selector_to_match):
                return error_signature, selector
    return None, None

def main():
    # Set up argument parser
    parser = argparse.ArgumentParser(description='Match Ethereum error selector with contract ABI.')
    parser.add_argument('abi_filename', type=str, help='The filename of the contract ABI JSON')
    parser.add_argument('error_selector', type=str, help='The error selector to match')

    args = parser.parse_args()

    # Ensure the error selector is in the correct format
    error_selector = args.error_selector.lower()

    # Load the ABI file
    with open(args.abi_filename, 'r') as abi_file:
        abi = json.load(abi_file)

    # Find matching error
    #print(f"looking for: {error_selector}")
    error_signature, selector = find_matching_error(abi, error_selector)

    if error_signature:
        print(f"Matched Error: {error_signature} -> {selector}")
    else:
        print(f"No matching error found for selector {args.error_selector}")

if __name__ == "__main__":
    main()

File: find_error.ts
#!/usr/bin/env -S deno run --allow-read

import { keccak_256 } from "npm:@noble/hashes/sha3";
import { bytesToHex } from "npm:@noble/hashes@1.6.1/utils";

/**
 * Generate a selector from a signature using keccak256
 * @param signature The function signature
 * @returns The selector (first 4 bytes of the hash)
 */
function generateSelector(signature: string): string {
  const encoder = new TextEncoder();
  const data = encoder.encode(signature);
  const hash = keccak_256(data);
  const hashHex = bytesToHex(hash);
  return "0x" + hashHex.substring(0, 8);
}

async function main() {
  // Check arguments
  if (Deno.args.length < 1) {
    console.error("Usage: find_error.ts <abi_filename> [0xerror_selector]");
    Deno.exit(1);
  }

  const abiFilename = Deno.args[0];
  let errorSelector = Deno.args.length > 1 ? Deno.args[1].toLowerCase() : null;
  
  // Ensure error selector starts with 0x if provided
  if (errorSelector && !errorSelector.startsWith("0x")) {
    errorSelector = "0x" + errorSelector;
  }

  try {
    // Read and parse the ABI file
    const abiContent = await Deno.readTextFile(abiFilename);
    const abi = JSON.parse(abiContent);

    // Extract all error types from the ABI
    const errors = abi.filter((item: any) => item.type === "error");
    
    // Process each error
    for (const error of errors) {
      const name = error.name;
      
      // Build the error signature by joining input types
      const inputs = error.inputs.map((input: any) => input.type).join(",");
      const errorSignature = `${name}(${inputs})`;
      
      // Generate the selector
      const selector = generateSelector(errorSignature);
      
      if (!errorSelector) {
        console.log(`Found error type: ${name} with selector: ${selector}`);
      }
      
      // Check if this selector matches the one we're looking for (if provided)
      if (errorSelector && selector.startsWith(errorSelector)) {
        console.log(`Matched Error: ${errorSignature} -> ${selector}`);
        Deno.exit(0);
      }
    }

    // If we get here and a selector was provided, no match was found
    if (errorSelector) {
      console.log(`No matching error found for selector ${errorSelector}`);
    }
  } catch (error) {
    console.error(`Error processing ABI file: ${error.message}`);
    Deno.exit(1);
  }
}

main();
